File: java/pdfclown.samples.cli/src/org/pdfclown/samples/cli/AcroFormParsingSample.java
package org.pdfclown.samples.cli;

import java.awt.geom.Rectangle2D;
import java.util.HashMap;
import java.util.Map;

import org.pdfclown.documents.Document;
import org.pdfclown.documents.Page;
import org.pdfclown.documents.interaction.annotations.Widget;
import org.pdfclown.documents.interaction.forms.Field;
import org.pdfclown.documents.interaction.forms.Form;
import org.pdfclown.files.File;
import org.pdfclown.util.io.IOUtils;

/**
  This sample demonstrates <b>how to inspect the AcroForm fields</b> of a PDF document.

  @author Stefano Chizzolini (http://www.stefanochizzolini.it)
  @since 0.0.7
  @version 0.2.0, 04/08/15
*/
public class AcroFormParsingSample
  extends Sample
{
  @Override
  public void run(
    )
  {
    File file = null;
    try
    {
      // 1. Opening the PDF file...
      {
        String filePath = promptFileChoice("Please select a PDF file");
        try
        {file = new File(filePath);}
        catch(Exception e)
        {throw new RuntimeException(filePath + " file access error.",e);}
      }
      Document document = file.getDocument();

      // 2. Get the acroform!
      Form form = document.getForm();
      if(!form.exists())
      {System.out.println("\nNo acroform available (AcroForm dictionary not found).");}
      else
      {
        System.out.println("\nIterating through the fields collection...\n");

        // 3. Showing the acroform fields...
        HashMap<String,Integer> objCounters = new HashMap<String,Integer>();
        for(Field field : form.getFields().values())
        {
          System.out.println("* Field '" + field.getFullName() + "' (" + field.getBaseObject() + ")");

          String typeName = field.getClass().getSimpleName();
          System.out.println("    Type: " + typeName);
          System.out.println("    Value: " + field.getValue());
          System.out.println("    Data: " + field.getBaseDataObject().toString());

          int widgetIndex = 0;
          for(Widget widget : field.getWidgets())
          {
            System.out.println("    Widget " + (++widgetIndex) + ":");
            Page widgetPage = widget.getPage();
            System.out.println("      Page: " + (widgetPage == null ? "undefined" : widgetPage.getNumber() + " (" + widgetPage.getBaseObject() + ")"));

            Rectangle2D widgetBox = widget.getBox();
            System.out.println("      Coordinates: {x:" + Math.round(widgetBox.getX()) + "; y:" + Math.round(widgetBox.getY()) + "; width:" + Math.round(widgetBox.getWidth()) + "; height:" + Math.round(widgetBox.getHeight()) + "}");
          }

          objCounters.put(typeName, (objCounters.containsKey(typeName) ? objCounters.get(typeName) : 0) + 1);
        }

        int fieldCount = form.getFields().size();
        if(fieldCount == 0)
        {System.out.println("No field available.");}
        else
        {
          System.out.println("\nFields partial counts (grouped by type):");
          for(Map.Entry<String,Integer> entry : objCounters.entrySet())
          {System.out.println(" " + entry.getKey() + ": " + entry.getValue());}
          System.out.println("Fields total count: " + fieldCount);
        }
      }
    }
    finally
    {
      // 4. Closing the PDF file...
      IOUtils.closeQuietly(file);
    }
  }
}

File: PdfClown-NetStandard/CliSamples/org/pdfclown/samples/cli/AcroFormParsingSample.cs
using org.pdfclown.documents;
using org.pdfclown.documents.contents.composition;
using org.pdfclown.documents.contents.entities;
using org.pdfclown.documents.contents.fonts;
using org.pdfclown.documents.contents.xObjects;
using org.pdfclown.documents.interaction.annotations;
using org.pdfclown.documents.interaction.forms;
using org.pdfclown.files;

using System;
using System.Collections.Generic;
using System.Drawing;

namespace org.pdfclown.samples.cli
{
  /**
    <summary>This sample demonstrates how to inspect the AcroForm fields of a PDF document.</summary>
  */
  public class AcroFormParsingSample
    : Sample
  {
    public override void Run(
      )
    {
      // 1. Opening the PDF file...
      string filePath = PromptFileChoice("Please select a PDF file");
      using(var file = new File(filePath))
      {
        Document document = file.Document;

        // 2. Get the acroform!
        Form form = document.Form;
        if(!form.Exists())
        {Console.WriteLine("\nNo acroform available (AcroForm dictionary not found).");}
        else
        {
          Console.WriteLine("\nIterating through the fields collection...\n");

          // 3. Showing the acroform fields...
          Dictionary<string,int> objCounters = new Dictionary<string,int>();
          foreach(Field field in form.Fields.Values)
          {
            Console.WriteLine("* Field '" + field.FullName + "' (" + field.BaseObject + ")");

            string typeName = field.GetType().Name;
            Console.WriteLine("    Type: " + typeName);
            Console.WriteLine("    Value: " + field.Value);
            Console.WriteLine("    Data: " + field.BaseDataObject.ToString());

            int widgetIndex = 0;
            foreach(Widget widget in field.Widgets)
            {
              Console.WriteLine("    Widget " + (++widgetIndex) + ":");
              Page widgetPage = widget.Page;
              Console.WriteLine("      Page: " + (widgetPage == null ? "undefined" : widgetPage.Number + " (" + widgetPage.BaseObject + ")"));

              RectangleF widgetBox = widget.Box;
              Console.WriteLine("      Coordinates: {x:" + Math.Round(widgetBox.X) + "; y:" + Math.Round(widgetBox.Y) + "; width:" + Math.Round(widgetBox.Width) + "; height:" + Math.Round(widgetBox.Height) + "}");
            }

            objCounters[typeName] = (objCounters.ContainsKey(typeName) ? objCounters[typeName] : 0) + 1;
          }

          int fieldCount = form.Fields.Count;
          if(fieldCount == 0)
          {Console.WriteLine("No field available.");}
          else
          {
            Console.WriteLine("\nFields partial counts (grouped by type):");
            foreach(KeyValuePair<string,int> entry in objCounters)
            {Console.WriteLine(" " + entry.Key + ": " + entry.Value);}
            Console.WriteLine("Fields total count: " + fieldCount);
          }
        }
      }
    }
  }
}